src/components/webview/MainWebviewProvider.ts

```diff
@@ -0,0 +1,501 @@
+import {
+	WebviewViewProvider,
+	WebviewView,
+	Uri,
+	ExtensionContext,
+	workspace,
+	commands,
+} from 'vscode';
+import { Message, MessageBus, MessageKind } from '../messageBus';
+import {
+	TreeNode,
+	View,
+	WebviewMessage,
+	WebviewResponse,
+} from './webviewEvents';
+import { WebviewResolver } from './WebviewResolver';
+import {
+	CaseElement,
+	Element,
+	ElementHash,
+	FileElement,
+	RootElement,
+} from '../../elements/types';
+import { Job, JobHash, JobKind } from '../../jobs/types';
+import { debounce } from '../../utilities';
+import { JobManager } from '../jobManager';
+import { CaseHash, CaseWithJobHashes } from '../../cases/types';
+import {
+	buildJobElement,
+	compareJobElements,
+} from '../../elements/buildJobElement';
+import {
+	buildFileElement,
+	compareFileElements,
+} from '../../elements/buildFileElement';
+import {
+	buildCaseElement,
+	compareCaseElements,
+} from '../../elements/buildCaseElement';
+import { CaseManager } from '../../cases/caseManager';
+
+const getElementIconBaseName = (kind: Element['kind']): string => {
+	switch (kind) {
+		case 'CASE':
+			return 'case.svg';
+		case 'FILE':
+			return 'ts2.svg';
+		default:
+			return 'bluelightbulb.svg';
+	}
+};
+
+export const ROOT_ELEMENT_HASH: ElementHash = '' as ElementHash;
+
+export class IntuitaProvider implements WebviewViewProvider {
+	__view: WebviewView | null = null;
+	__extensionPath: Uri;
+	__webviewResolver: WebviewResolver | null = null;
+	__elementMap = new Map<ElementHash, Element>();
+
+	constructor(
+		context: ExtensionContext,
+		private readonly __messageBus: MessageBus,
+		private readonly __jobManager: JobManager,
+		private readonly __caseManager: CaseManager,
+	) {
+		this.__extensionPath = context.extensionUri;
+
+		this.__webviewResolver = new WebviewResolver(this.__extensionPath);
+	}
+
+	refresh(): void {
+		if (!this.__view) {
+			return;
+		}
+
+		this.__webviewResolver?.resolveWebview(this.__view.webview, 'main', {});
+	}
+
+	resolveWebviewView(webviewView: WebviewView): void | Thenable<void> {
+		if (!webviewView.webview) return;
+
+		this.__webviewResolver?.resolveWebview(webviewView.webview, 'main', {});
+		this.__view = webviewView;
+
+		this.__view.onDidChangeVisibility(() => {
+			this.__onUpdateElementsMessage();
+		});
+
+		this.__attachExtensionEventListeners();
+		this.__attachWebviewEventListeners();
+	}
+
+	public setView(data: View) {
+		this.__postMessage({
+			kind: 'webview.global.setView',
+			value: data,
+		});
+	}
+
+	private __postMessage(message: WebviewMessage) {
+		if (!this.__view) {
+			return;
+		}
+
+		this.__view.webview.postMessage(message);
+	}
+
+	private __getTree = (element: Element): TreeNode => {
+		const mappedNode: TreeNode = {
+			id: element.hash,
+		};
+
+		mappedNode.label = 'label' in element ? element.label : 'Recipe';
```

Comment:
is there an element of kind `RECIPE`?

Reply:
I will make an update later so ROOT element is the Recipe and has a label with recipe name